In [1]:
import pandas as pd
import numpy as np

In [176]:
ds = pd.read_csv(r'C:\Users\TEMP\Downloads\titanic.csv')
ds.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [177]:
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [178]:
len(ds.columns)

12

In [179]:
ds.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [180]:
ds.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [181]:
ds.shape

(891, 12)

### loc, iloc

In [182]:
ds.iloc[0:3]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [183]:
ds.iloc[0:5, 0:3]

PassengerId  Survived  Pclass
0            1         0       3
1            2         1       1
2            3         1       3
3            4         1       1
4            5         0       3

In [184]:
ds.iloc[[0, 11, 88], [0, -1]]

PassengerId Embarked
0             1        S
11           12        S
88           89        S

In [185]:
# заменим порядковые индексы на имена пассажиров
names = ds['Name']
ds_with_names = ds.copy(deep=True) # без 'copy' данные этих датафрэймов будут зависимыми
ds_with_names.index = list(names.values)
ds_with_names.head()

PassengerId  Survived  \
Braund, Mr. Owen Harris                                       1         0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...            2         1   
Heikkinen, Miss. Laina                                        3         1   
Futrelle, Mrs. Jacques Heath (Lily May Peel)                  4         1   
Allen, Mr. William Henry                                      5         0   

                                                    Pclass  \
Braund, Mr. Owen Harris                                  3   
Cumings, Mrs. John Bradley (Florence Briggs Tha...       1   
Heikkinen, Miss. Laina                                   3   
Futrelle, Mrs. Jacques Heath (Lily May Peel)             1   
Allen, Mr. William Henry                                 3   

                                                                                                 Name  \
Braund, Mr. Owen Harris                                                       Braund, Mr. Owen Harris   
Cumings, Mrs. John Bradley (Florence Briggs Tha...  Cumings, Mrs. John Bradley (Florence Briggs Th...   
Heikkinen, Miss. Laina                                                         Heikkinen, Miss. Laina   
Futrelle, Mrs. Jacques Heath (Lily May Peel)             Futrelle, Mrs. Jacques Heath (Lily May Peel)   
Allen, Mr. William Henry                                                     Allen, Mr. William Henry   

                                                       Sex   Age  SibSp  \
Braund, Mr. Owen Harris                               male  22.0      1   
Cumings, Mrs. John Bradley (Florence Briggs Tha...  female  38.0      1   
Heikkinen, Miss. Laina                              female  26.0      0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)        female  35.0      1   
Allen, Mr. William Henry                              male  35.0      0   

                                                    Parch            Ticket  \
Braund, Mr. Owen Harris                                 0         A/5 21171   
Cumings, Mrs. John Bradley (Florence Briggs Tha...      0          PC 17599   
Heikkinen, Miss. Laina                                  0  STON/O2. 3101282   
Futrelle, Mrs. Jacques Heath (Lily May Peel)            0            113803   
Allen, Mr. William Henry                                0            373450   

                                                       Fare Cabin Embarked  
Braund, Mr. Owen Harris                              7.2500   NaN        S  
Cumings, Mrs. John Bradley (Florence Briggs Tha...  71.2833   C85        C  
Heikkinen, Miss. Laina                               7.9250   NaN        S  
Futrelle, Mrs. Jacques Heath (Lily May Peel)        53.1000  C123        S  
Allen, Mr. William Henry                             8.0500   NaN        S

In [186]:
ds_with_names.loc[['Heikkinen, Miss. Laina', 'Dooley, Mr. Patrick'], ['Sex', 'Age']]

Sex   Age
Heikkinen, Miss. Laina  female  26.0
Dooley, Mr. Patrick       male  32.0

In [187]:
# Датафрэйм - это объединенные пандовские серии!!!
pd_series_1 = pd.Series(["Саша", "Паша", "Миша"])
pd_series_2 = pd.Series(["Саша", "Маша", "Настя"])
df = pd.DataFrame({'Boys': pd_series_1, 'Girls': pd_series_2})
df

Boys  Girls
0  Саша   Саша
1  Паша   Маша
2  Миша  Настя

In [188]:
# Обращение как к пандовской серии
ds['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [189]:
# Обращение как к датафрэйму
ds[['Age']]

Age
0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
..    ...
886  27.0
887  19.0
888   NaN
889  26.0
890  32.0

[891 rows x 1 columns]

# Фильтрация данных

In [21]:
ds = pd.read_csv(r'C:\Users\TEMP\Downloads\StudentsPerformance.csv')
ds.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [207]:
reduce_lunch = ds.loc[ds.lunch == 'free/reduced']
res = len(reduce_lunch) / len(ds)
res

0.355

In [208]:
standard_lunch = ds.loc[ds['lunch'] == 'standard'] 

In [209]:
reduce_lunch.describe()

math score  reading score  writing score
count  355.000000     355.000000     355.000000
mean    58.921127      64.653521      63.022535
std     15.159956      14.895339      15.433823
min      0.000000      17.000000      10.000000
25%     49.000000      56.000000      53.000000
50%     60.000000      65.000000      64.000000
75%     69.000000      75.000000      74.000000
max    100.000000     100.000000     100.000000

In [210]:
standard_lunch.describe()

math score  reading score  writing score
count  645.000000     645.000000     645.000000
mean    70.034109      71.654264      70.823256
std     13.653501      13.830602      14.339487
min     19.000000      26.000000      22.000000
25%     61.000000      63.000000      62.000000
50%     69.000000      72.000000      72.000000
75%     80.000000      82.000000      81.000000
max    100.000000     100.000000     100.000000

### query

In [22]:
ds = ds.rename(columns={'math score':'math_score', 'reading score':'reading_score', 'writing score':'writing_score'})
ds

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math_score  reading_score  writing_score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

In [218]:
q1 = ds.query("math_score > 80")

In [223]:
q2 = ds.query("math_score > 99 & reading_score > 99 & writing_score > 99")
q2
# обязательно '&' и '|' вместо 'and' и 'or', так как работаем с пандовскими сериями, а не с 2 булевыми значениями!!!

gender race/ethnicity parental level of education     lunch  \
458  female        group E           bachelor's degree  standard   
916    male        group E           bachelor's degree  standard   
962  female        group E          associate's degree  standard   

    test preparation course  math_score  reading_score  writing_score  
458                    none         100            100            100  
916               completed         100            100            100  
962                    none         100            100            100

# Сортировка

In [33]:
# по одному признаку по убыванию
ds.sort_values('race/ethnicity', ascending=False)

gender race/ethnicity parental level of education         lunch  \
676  female        group E                some college      standard   
638    male        group E                some college      standard   
689    male        group E                some college  free/reduced   
686    male        group E                some college      standard   
685  female        group E             master's degree      standard   
..      ...            ...                         ...           ...   
688    male        group A                 high school  free/reduced   
151    male        group A           bachelor's degree      standard   
506    male        group A                 high school      standard   
511    male        group A            some high school      standard   
62     male        group A          associate's degree  free/reduced   

    test preparation course  math_score  reading_score  writing_score  
676               completed          73             78             76  
638                    none          86             76             74  
689                    none          93             90             83  
686               completed          85             75             68  
685               completed          94             99            100  
..                      ...         ...            ...            ...  
688                    none          53             58             44  
151                    none          77             67             68  
506                    none          68             70             66  
511                    none          64             50             43  
62                     none          62             61             55  

[1000 rows x 8 columns]

In [34]:
# по нескольким признакам в разном порядке
ds.sort_values(['math_score', 'reading_score'], ascending=[False, True]).head()

gender race/ethnicity parental level of education         lunch  \
451  female        group E                some college      standard   
623    male        group A                some college      standard   
625    male        group D                some college      standard   
149    male        group E          associate's degree  free/reduced   
458  female        group E           bachelor's degree      standard   

    test preparation course  math_score  reading_score  writing_score  
451                    none         100             92             97  
623               completed         100             96             86  
625               completed         100             97             99  
149               completed         100            100             93  
458                    none         100            100            100

# Группировка и аггрегация 
### groupby, aggregate

In [54]:
ds = pd.read_csv("StudentsPerformance.csv")
ds.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [15]:
# группировка по полу без подписей по признакам
ds.groupby('gender').mean()

math score  reading score  writing score
gender                                          
female   63.633205      72.608108      72.467181
male     68.728216      65.473029      63.311203

In [28]:
# отключить переназначение индексов
ds.groupby('gender', as_index=False).mean()

gender  math score  reading score  writing score
0  female   63.633205      72.608108      72.467181
1    male   68.728216      65.473029      63.311203

In [16]:
# группировка по полу с подписями по признакам (mean)
ds.groupby('gender').aggregate(['mean'])

math score reading score writing score
             mean          mean          mean
gender                                       
female  63.633205     72.608108     72.467181
male    68.728216     65.473029     63.311203

In [25]:
# группировка по условным расам и аггрегация 
ds.groupby('race/ethnicity').aggregate(['mean', 'std'])

math score            reading score            writing score  \
                     mean        std          mean        std          mean   
race/ethnicity                                                                
group A         61.629213  14.523008     64.674157  15.543762     62.674157   
group B         63.452632  15.468191     67.352632  15.177499     65.600000   
group C         64.463950  14.852666     69.103448  13.997033     67.827586   
group D         67.362595  13.769386     70.030534  13.895306     70.145038   
group E         73.821429  15.534259     73.028571  14.874024     71.407143   

                           
                      std  
race/ethnicity             
group A         15.468278  
group B         15.625173  
group C         14.983378  
group D         14.367707  
group E         15.113906

In [33]:
# группировка по условным расам и агрегация некоторых числовых признаков с дальнейшим их переименованием
ds.groupby('race/ethnicity').aggregate({'math score':'mean', 'reading score':'std'})\
.rename(columns={'math score': 'mean_math_score', 'reading score':'std_reading_score'})

mean_math_score  std_reading_score
race/ethnicity                                    
group A               61.629213          15.543762
group B               63.452632          15.177499
group C               64.463950          13.997033
group D               67.362595          13.895306
group E               73.821429          14.874024

## Группировка по нескольким индексам

In [37]:
ds.groupby(['gender', 'race/ethnicity']).mean()
# вознакет мультииндекс - неудобно

math score  reading score  writing score
gender race/ethnicity                                          
female group A          58.527778      69.000000      67.861111
       group B          61.403846      71.076923      70.048077
       group C          62.033333      71.944444      71.777778
       group D          65.248062      74.046512      75.023256
       group E          70.811594      75.840580      75.536232
male   group A          63.735849      61.735849      59.150943
       group B          65.930233      62.848837      60.220930
       group C          67.611511      65.424460      62.712230
       group D          69.413534      66.135338      65.413534
       group E          76.746479      70.295775      67.394366

In [42]:
# убираем мультииндекс
ds.groupby(['gender', 'race/ethnicity'], as_index=False).mean()

gender race/ethnicity  math score  reading score  writing score
0  female        group A   58.527778      69.000000      67.861111
1  female        group B   61.403846      71.076923      70.048077
2  female        group C   62.033333      71.944444      71.777778
3  female        group D   65.248062      74.046512      75.023256
4  female        group E   70.811594      75.840580      75.536232
5    male        group A   63.735849      61.735849      59.150943
6    male        group B   65.930233      62.848837      60.220930
7    male        group C   67.611511      65.424460      62.712230
8    male        group D   69.413534      66.135338      65.413534
9    male        group E   76.746479      70.295775      67.394366

In [44]:
gender_scores = ds.groupby(['gender', 'race/ethnicity']).mean()
# отберем сстроки по составному индексу (мультииндексу)
gender_scores.loc[('male', 'group E')]

math score       76.746479
reading score    70.295775
writing score    67.394366
Name: (male, group E), dtype: float64

In [46]:
# или несколько строк
gender_scores.loc[[('male', 'group A'), ('female', 'group A')]]

math score  reading score  writing score
gender race/ethnicity                                          
male   group A          63.735849      61.735849      59.150943
female group A          58.527778      69.000000      67.861111

Найдем число уникальных оценок среди студентов одинакового пола и расы:

In [64]:
ds.rename(columns = {'math score':'math_score'}, inplace=True)
ds.groupby(['gender', 'race/ethnicity']).math_score.nunique()

gender  race/ethnicity
female  group A           29
        group B           51
        group C           59
        group D           53
        group E           44
male    group A           38
        group B           43
        group C           56
        group D           49
        group E           38
Name: math_score, dtype: int64

Найдем топ-5 студентов среди студентов разного пола:

In [77]:
ds.sort_values(['gender', 'math_score'], ascending=False).groupby('gender').head()

gender race/ethnicity parental level of education         lunch  \
149    male        group E          associate's degree  free/reduced   
623    male        group A                some college      standard   
625    male        group D                some college      standard   
916    male        group E           bachelor's degree      standard   
306    male        group E                some college      standard   
451  female        group E                some college      standard   
458  female        group E           bachelor's degree      standard   
962  female        group E          associate's degree      standard   
114  female        group E           bachelor's degree      standard   
263  female        group E                 high school      standard   

    test preparation course  math_score  reading score  writing score  
149               completed         100            100             93  
623               completed         100             96             86  
625               completed         100             97             99  
916               completed         100            100            100  
306               completed          99             87             81  
451                    none         100             92             97  
458                    none         100            100            100  
962                    none         100            100            100  
114               completed          99            100            100  
263                    none          99             93             90

## Добавление и удаление колонок

Добавим суммарный балл по всем 3 предметам: 

In [80]:
ds.rename(columns={'reading score': 'reading_score', 'writing score': 'writing_score'}, inplace=True)
total = ds.math_score + ds.reading_score + ds.writing_score
ds['total_score'] = total
ds.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score  
0          218  
1          247  
2          278  
3          148  
4          229

In [83]:
ds['total_score_log'] = np.log(total)
ds.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score  total_score_log  
0          218         5.384495  
1          247         5.509388  
2          278         5.627621  
3          148         4.997212  
4          229         5.433722

### assign

In [92]:
ds.assign(total_score_sin = np.sin(ds.total_score)).head()
# не изменяет исходный датафрэйм!

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score  total_score_log  total_score_sin  
0          218         5.384495        -0.942525  
1          247         5.509388         0.926807  
2          278         5.627621         0.999521  
3          148         4.997212        -0.338333  
4          229         5.433722         0.329962

In [94]:
ds.head()
# total_score_sin - отсутствует

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score  total_score_log  
0          218         5.384495  
1          247         5.509388  
2          278         5.627621  
3          148         4.997212  
4          229         5.433722

Удалим все оценки кроме суммарных:

In [99]:
ds.drop(['math_score', 'reading_score', 'writing_score', 'total_score_log'], axis=1).head()
# axis=1 - смотрим на названия колонок, 
# по умолчанию axis=0 - индексы строк

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  total_score  
0                    none          218  
1               completed          247  
2                    none          278  
3                    none          148  
4                    none          229

In [91]:
ds

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math_score  reading_score  writing_score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

     total_score  total_score_log  
0            218         5.384495  
1            247         5.509388  
2            278         5.627621  
3            148         4.997212  
4            229         5.433722  
..           ...              ...  
995          282         5.641907  
996          172         5.147494  
997          195         5.273000  
998          223         5.407172  
999          249         5.517453  

[1000 rows x 10 columns]

Рассчитаем количество героев, группировав их по количеству ног:

In [132]:
df = pd.read_csv('dota_hero_stats.csv')
df.head()

Unnamed: 0 attack_type  id  legs  localized_name  \
0           0       Melee   1     2       Anti-Mage   
1           1       Melee   2     2             Axe   
2           2      Ranged   3     4            Bane   
3           3       Melee   4     2     Bloodseeker   
4           4      Ranged   5     2  Crystal Maiden   

                           name primary_attr  \
0        npc_dota_hero_antimage          agi   
1             npc_dota_hero_axe          str   
2            npc_dota_hero_bane          int   
3     npc_dota_hero_bloodseeker          agi   
4  npc_dota_hero_crystal_maiden          int   

                                               roles  
0                       ['Carry', 'Escape', 'Nuker']  
1    ['Initiator', 'Durable', 'Disabler', 'Jungler']  
2        ['Support', 'Disabler', 'Nuker', 'Durable']  
3  ['Carry', 'Disabler', 'Jungler', 'Nuker', 'Ini...  
4        ['Support', 'Disabler', 'Nuker', 'Jungler']

In [140]:
df.groupby('legs').count()

Unnamed: 0  attack_type  id  localized_name  name  primary_attr  roles
legs                                                                        
0             11           11  11              11    11            11     11
2             95           95  95              95    95            95     95
4              7            7   7               7     7             7      7
6              3            3   3               3     3             3      3
8              1            1   1               1     1             1      1

In [120]:
df['legs'].value_counts()

2    95
0    11
4     7
6     3
8     1
Name: legs, dtype: int64

Найдем наиболее встречающийся набор характеристик:

In [168]:
df.groupby(['attack_type', 'primary_attr']).nunique()
# 40 - Ranged int

Unnamed: 0  attack_type  id  legs  localized_name  \
attack_type primary_attr                                                      
Melee       agi                   19            1  19     4              19   
            int                    2            1   2     1               2   
            str                   35            1  35     4              35   
Ranged      agi                   18            1  18     3              18   
            int                   40            1  40     4              40   
            str                    3            1   3     2               3   

                          name  primary_attr  roles  
attack_type primary_attr                             
Melee       agi             19             1     19  
            int              2             1      2  
            str             35             1     31  
Ranged      agi             18             1     16  
            int             40             1     33  
            str              3             1      3

In [139]:
df.groupby(['attack_type', 'primary_attr']).count()

Unnamed: 0  id  legs  localized_name  name  roles
attack_type primary_attr                                                   
Melee       agi                   19  19    19              19    19     19
            int                    2   2     2               2     2      2
            str                   35  35    35              35    35     35
Ranged      agi                   18  18    18              18    18     18
            int                   40  40    40              40    40     40
            str                    3   3     3               3     3      3

К нам поступили данные из бухгалтерии о заработках Лупы и Пупы за разные задачи! Посмотрим у кого из них больше средний заработок в различных категориях (колонка Type):

In [129]:
df = pd.read_csv('accountancy.csv')
df.head()

Unnamed: 0 Executor Type  Salary
0           0     Pupa    D      63
1           1     Pupa    A     158
2           2     Pupa    D     194
3           3     Pupa    E     109
4           4    Loopa    E     184

In [131]:
df.groupby(['Executor', 'Type']).max()

Unnamed: 0  Salary
Executor Type                    
Loopa    A             46      58
         B             43     249
         C             47     249
         D             49     181
         E             42     220
         F             35     244
Pupa     A             39     234
         B             41      77
         C             24      94
         D             31     194
         E             30     197
         F             48     195

Аспирант Ростислав изучает метаболом водорослей и получил такую табличку. В ней он записал вид каждой водоросли, её род (группа, объединяющая близкие виды), группа (ещё одно объединение водорослей в крупные фракции) и концентрации анализируемых веществ:

In [141]:
ds = pd.read_csv('algae.csv')
ds.head()

species       genus  group   sucrose    alanin   citrate  \
0    Fucus_vesiculosus       Fucus  brown  3.001472  3.711498  5.004262   
1  Saccharina_japonica  Saccharina  brown  6.731070  1.255251  5.621499   
2       Fucus_serratus       Fucus  brown  3.276870  0.346431  1.216767   
3      Fucus_distichus       Fucus  brown  6.786996  6.641303  6.423606   
4    Cladophora_fracta  Cladophora  green  3.861470  1.648450  6.940588   

    glucose  oleic_acid  
0  2.548459    6.405165  
1  6.013219    4.156700  
2  3.623225    0.304573  
3  2.272724    3.393203  
4  2.316955    2.528886

Найдем среднюю концентрацию каждого из веществ в каждом из родов (колонка genus):

In [143]:
mean_concentrations = ds.groupby('genus').mean()
mean_concentrations

sucrose    alanin   citrate   glucose  oleic_acid
genus                                                          
Ascophyllum  6.825467  0.875429  5.253527  3.414961    2.432526
Cladophora   4.008792  3.997055  5.288311  2.800276    2.263472
Fucus        4.355112  3.566411  4.214878  2.814803    3.367647
Palmaria     0.704580  3.176440  5.573905  3.242090    2.245538
Saccharina   4.183596  3.524207  3.346710  4.980594    4.487252

Рассчитаем чему равна минимальная, средняя и максимальная концентрации аланина (alanin) среди видов рода Fucus:

In [152]:
ds.groupby('genus').aggregate(['min', 'mean', 'max']).alanin

min      mean       max
genus                                    
Ascophyllum  0.875429  0.875429  0.875429
Cladophora   1.648450  3.997055  6.881024
Fucus        0.346431  3.566411  6.641303
Palmaria     3.176440  3.176440  3.176440
Saccharina   1.255251  3.524207  5.793163

Сгруппируем данные по переменной group и ответим на вопросы:
1. Размах (максимум - минимум) сахарозы в бурых, зелёных и красных водорослях
2. Число видов бурых, зелёных и красных водорослей
3. Дисперсия концентрации цитрата в бурых, зелёных и красных водорослях

In [162]:
ds.groupby('group').describe().sucrose

count      mean       std       min       25%      50%       75%  \
group                                                                     
brown    6.0  4.709666  2.336471  1.636122  3.070321  5.00397  6.773014   
green    3.0  4.008792  1.634817  2.452623  3.157047  3.86147  4.786877   
red      1.0  0.704580       NaN  0.704580  0.704580  0.70458  0.704580   

            max  
group            
brown  6.825467  
green  5.712284  
red    0.704580

Проанализируем набор данных по оттоку клиентов телеком-оператора

In [2]:
ds = pd.read_csv("telecom_churn.csv")
ds.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [141]:
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

In [142]:
ds.shape

(3333, 21)

Заменим булевый тип 'churn' на целочисленный:

In [10]:
ds['churn'] = ds['churn'].astype('int')

Посмотрим, сколько человек уходят, а сколько остаются лояльными:

In [144]:
ds['churn'].value_counts()

0    2850
1     483
Name: churn, dtype: int64

Видим, что большинство лояльны, но значительная часть оттекает.  
Нормализуем полученные значения:

In [145]:
ds['churn'].value_counts(normalize=True)

0    0.855086
1    0.144914
Name: churn, dtype: float64

Таким образом, если бы мы относили всех пользователей к лояльным, то мы бы ошибались в 14,4914% раз, соответсвенно показатель accuracy = 0.855086. Это значение будем считать базовым для анализа (baseline). Дальнейшие исследования должны предлагать решения, обеспечивающие accuracy > 0.855086.  
Сгруппируем данные по лояльности клиентов:

In [11]:
ds.groupby('churn').mean()

account length   area code  number vmail messages  total day minutes  \
churn                                                                         
0          100.793684  437.074737               8.604561         175.175754   
1          102.664596  437.817805               5.115942         206.914079   

       total day calls  total day charge  total eve minutes  total eve calls  \
churn                                                                          
0           100.283158         29.780421         199.043298       100.038596   
1           101.335404         35.175921         212.410145       100.561077   

       total eve charge  total night minutes  total night calls  \
churn                                                             
0             16.918909           200.133193         100.058246   
1             18.054969           205.231677         100.399586   

       total night charge  total intl minutes  total intl calls  \
churn                                                             
0                9.006074           10.158877          4.532982   
1                9.235528           10.700000          4.163561   

       total intl charge  customer service calls  
churn                                             
0               2.743404                1.449825  
1               2.889545                2.229814

In [12]:
ds.pivot_table(['total intl calls', 'total intl charge', 'customer service calls'], ['churn'], aggfunc='mean')

customer service calls  total intl calls  total intl charge
churn                                                             
0                    1.449825          4.532982           2.743404
1                    2.229814          4.163561           2.889545

Уже сейчас видно, что НЕлояльные пользователи чаще звонят в сервисную службу ('customer service calls') и платят больше в роуминге ('total intl charge') за меньшее количество звонков ('total intl calls'). 
  
Рассмотрим как наблюдения в нашей выборке распределены в контексте двух признаков — 'сhurn' и 'international plan'. Для этого построим таблицу сопряженности, воспользовавшись методом *crosstab*:

In [147]:
pd.crosstab(ds['churn'], ds['international plan'])

international plan    no  yes
churn                        
0                   2664  186
1                    346  137

Нормализуем значения:

In [148]:
pd.crosstab(ds['churn'], ds['international plan'], normalize=True)

international plan       no       yes
churn                                
0                   0.79928  0.055806
1                   0.10381  0.041104

In [149]:
pd.crosstab(ds['churn'], ds['voice mail plan'], normalize=True)

voice mail plan        no       yes
churn                              
0                0.602460  0.252625
1                0.120912  0.024002

Таким образом, видно, что большинство пользователей лояльны и при этом пользуются дополнительными услугами (международного роуминга / голосовой почты), но при этом доля оттока выше у клиентов с подключенным роумингом. Это может означать, что клиенты недовольны неожиданно высокими платами за разговор в роуминге.

In [3]:
pd.crosstab(ds['churn'], ds['international plan'], margins=True)

international plan    no  yes   All
churn                              
False               2664  186  2850
True                 346  137   483
All                 3010  323  3333

Далее посмотрим на еще один важный признак – "Число обращений в сервисный центр" ('сustomer service calls'). Также построим сводную таблицу:

In [151]:
pd.crosstab(ds['churn'], ds['customer service calls'])

customer service calls    0     1    2    3   4   5   6  7  8  9
churn                                                           
0                       605  1059  672  385  90  26   8  4  1  0
1                        92   122   87   44  76  40  14  5  1  2

Видим, что начиная с 4 звонка, доля оттока клиентов сильно возрастает.  
Добавим еще одну "фичу" булевого типа - много ли раз (>3) клиент звонит в сервис:

In [152]:
ds['many_calls'] = (ds['customer service calls'] >= 4).astype(int)
ds['many_calls']

0       0
1       0
2       0
3       0
4       0
       ..
3328    0
3329    0
3330    0
3331    0
3332    0
Name: many_calls, Length: 3333, dtype: int32

In [153]:
pd.crosstab(ds['many_calls'], ds['churn'])

churn          0    1
many_calls           
0           2721  345
1            129  138

Видим, что доля в этом случае клиент скорее всего уйдет

In [154]:
d = {'yes': True, 'no': False}
ds['international plan'] = ds['international plan'].map(d)

In [160]:
pd.crosstab(ds['many_calls'] & ds['international plan'], ds['churn'])

churn     0    1
row_0           
False  2841  464
True      9   19

In [159]:
pd.crosstab(ds['many_calls'] & ds['international plan'], ds['churn'], normalize=True)

churn         0         1
row_0                    
False  0.852385  0.139214
True   0.002700  0.005701

Значит, прогнозируя отток клиента в случае, когда число звонков в сервисный центр больше 3 
и подключен роуминг (и прогнозируя лояльность – в противном случае), можно ожидать около 85.8% 
правильных попаданий (accuracy).  
Эти 85.8% > 85.5%, 
которые мы получили с помощью тривиального классификатора (когда считали, что все клиенты лояльны), это новая неплохая отправная точка (baseline) 
для дальнейших моделей машинного обучения.

# Задания из курса Stepik
**1.**  
Любым удобным для вас способом создайте dataframe c именем my_data, в котором две колонки c именами (type - строки, value - целые числа) и четыре наблюдения в каждой колонке:    
<table>
<tr>
<td>type</td>
<td>value</td>
</tr>
<tr>
<td>A</td>
<td>10</td>    
</tr>
<tr>
<td>A</td>
<td>14</td>    
</tr>
<tr>
<td>B</td>
<td>12</td>    
</tr>
<tr>
<td>B</td>
<td>23</td>    
</tr>
</table>

In [2]:
import pandas as pd
cols = ['type', 'value']
data = [['A', 10], ['A', 14], ['B', 12], ['B', 23]]
my_data = pd.DataFrame(data, columns=cols)
my_data

type  value
0    A     10
1    A     14
2    B     12
3    B     23

In [5]:
data = pd.read_clipboard()
my_data = pd.DataFrame(data)
my_data

type  value
0    A     10
1    A     14
2    B     12
3    B     23

**2.**  
1) В переменную с именем subset_1 сохраните только первые 10 строк и только 1 и 3 колонку.  
2) В переменную с именем subset_2 сохраните все строки кроме 1 и 5 и только 2 и 4 колонку.

In [27]:
ds = pd.read_csv('my_stat.csv')
ds.head()

V1    V2 V3   V4
0   2  13.0  B  1.0
1  -1   7.0  B  1.0
2   0  11.0  A  0.0
3   2  11.0  A  2.0
4   0  10.0  B -1.0

In [42]:
# 1)
subset_1 = ds.iloc[:10, [0, 2]]
subset_1

V1 V3
0   2  B
1  -1  B
2   0  A
3   2  A
4   0  B
5   2  A
6   0  A
7   1  A
8   0  B
9  -1  A

In [43]:
# 2)
subset_2  = ds.index.isin([0, 5])
subset_2 = ds[~s].iloc[:, [1, 3]]
subset_2

V2   V4
1     7.0  1.0
2    11.0  0.0
3    11.0  2.0
4    10.0 -1.0
6     9.0  2.0
..    ...  ...
495   8.0 -1.0
496   8.0  1.0
497  11.0  2.0
498  11.0  2.0
499   8.0  1.0

[498 rows x 2 columns]

In [45]:
# 2)
subset_2 = ds.iloc[:, [1, 3]]
subset_2 = subset_2.drop([0, 4])
subset_2

V2   V4
1     7.0  1.0
2    11.0  0.0
3    11.0  2.0
5    11.0  0.0
6     9.0  2.0
..    ...  ...
495   8.0 -1.0
496   8.0  1.0
497  11.0  2.0
498  11.0  2.0
499   8.0  1.0

[498 rows x 2 columns]

**3.**  
В переменную subset_1 сохраните только те наблюдения, у которых значения переменной V1  строго больше 0, и значение переменной V3  равняется 'A'.

В переменную  subset_2  сохраните только те наблюдения, у которых значения переменной V2  не равняются 10, или значения переменной V4 больше или равно 1.

In [67]:
ds = pd.read_csv('my_stat.csv')
ds.head()

V1    V2 V3   V4
0   2  13.0  B  1.0
1  -1   7.0  B  1.0
2   0  11.0  A  0.0
3   2  11.0  A  2.0
4   0  10.0  B -1.0

In [62]:
subset_1 = ds.query("(V1 > 0) & (V3 == 'A')")
subset_1

V1    V2 V3   V4
3     2  11.0  A  2.0
5     2  11.0  A  0.0
7     1   9.0  A  1.0
10    1  10.0  A  1.0
13    1  12.0  A  3.0
..   ..   ... ..  ...
491   4   7.0  A  2.0
494   1  11.0  A  1.0
495   1   8.0  A -1.0
496   3   8.0  A  1.0
498   1  11.0  A  2.0

[88 rows x 4 columns]

In [63]:
subset_2 = ds.query("(V2 != 10) | (V4 >= 1)")
subset_2

V1    V2 V3   V4
0     2  13.0  B  1.0
1    -1   7.0  B  1.0
2     0  11.0  A  0.0
3     2  11.0  A  2.0
5     2  11.0  A  0.0
..   ..   ... ..  ...
495   1   8.0  A -1.0
496   3   8.0  A  1.0
497  -1  11.0  B  2.0
498   1  11.0  A  2.0
499   0   8.0  B  1.0

[447 rows x 4 columns]

**4.**  
В переменной *ds* лежат данные с которыми вам необходимо проделать следующее дейтвие. В этих данных (*ds*) создайте две новые переменных:  
*V5 = V1 + V4*  
*V6* = натуральный логарифм переменной *V2*

In [69]:
import numpy as np
ds['V5'] = ds['V1'] + ds['V4']
ds['V6'] = np.log(ds['V2'])
ds

V1    V2 V3   V4   V5        V6
0     2  13.0  B  1.0  3.0  2.564949
1    -1   7.0  B  1.0  0.0  1.945910
2     0  11.0  A  0.0  0.0  2.397895
3     2  11.0  A  2.0  4.0  2.397895
4     0  10.0  B -1.0 -1.0  2.302585
..   ..   ... ..  ...  ...       ...
495   1   8.0  A -1.0  0.0  2.079442
496   3   8.0  A  1.0  4.0  2.079442
497  -1  11.0  B  2.0  1.0  2.397895
498   1  11.0  A  2.0  3.0  2.397895
499   0   8.0  B  1.0  1.0  2.079442

[500 rows x 6 columns]

**5.**   
Переменные V1, V2  ... такие имена никуда не годятся. С такими названиями легко запутаться в собственных данных и в результате ошибиться в расчетах.  
Переименуйте колонки в данных  *ds* следующим образом:  
*V1 -> session_value  
V2 -> group  
V3 -> time  
V4 -> n_users*

In [81]:
ds = pd.read_csv('my_stat.csv')
ds = ds.rename(columns={'V1':'session_value', 'V2':'group', 'V3':'time', 'V4':'n_users'})
ds.head()

session_value  group time  n_users
0              2   13.0    B      1.0
1             -1    7.0    B      1.0
2              0   11.0    A      0.0
3              2   11.0    A      2.0
4              0   10.0    B     -1.0

**6.**  
В dataframe с именем *ds* сохранены данные с 4 колонками: *session_value, group, time, n_users*.  
В переменной *session_value* замените все пропущенные значения на нули.  
В переменной *n_users* замените все отрицательные значения на медианное значение переменной *n_users* (без учета отрицательных значений, разумеется).

In [73]:
# заполняем NaN нулями
ds['session_value'].fillna(0) 
# выбираем неотрицательные значения, чтобы по ним рассчитать медиану 
positive = ds.query("n_users >= 0")
# рассчитываем медиану
m = positive.median()
# меняем отрицательные значения на медианные
ds.loc[(ds.n_users < 0), 'n_users'] = m.n_users

session_value  group time  n_users
0                2   13.0    B      1.0
1               -1    7.0    B      1.0
2                0   11.0    A      0.0
3                2   11.0    A      2.0
4                0   10.0    B     -1.0
..             ...    ...  ...      ...
495              1    8.0    A     -1.0
496              3    8.0    A      1.0
497             -1   11.0    B      2.0
498              1   11.0    A      2.0
499              0    8.0    B      1.0

[500 rows x 4 columns]

In [ ]:
# еще 1 способ заменить отрицательные значения на медианные 
ds.loc[ds['n_users'] < 0, 'n_users'] = ds.loc[ds['n_users'] >= 0, 'n_users'].median()

**7.**  
В этой задаче для данных *ds* рассчитайте среднее значение переменной *session_value* для каждой группы (переменная *group*), в получившемся dataframe  переменная *group* не должна превратиться в индекс. Также переименуйте колонку со средним значением *session_value* в *mean_session_value*.

In [117]:
res = ds.groupby('group', as_index=False).mean().rename(columns={'session_value':'mean_session_value'})
res = res.loc[:, 'group':'mean_session_value']
res.head()

group  mean_session_value
0    5.0            0.000000
1    7.0            0.333333
2    8.0            0.633333
3    9.0            0.238938
4   10.0            0.065990

In [118]:
res = ds.groupby('group', as_index=False)\
.agg({'session_value':'mean'}).rename(columns={'session_value':'mean_session_value'})
res.head()

group  mean_session_value
0    5.0            0.000000
1    7.0            0.333333
2    8.0            0.633333
3    9.0            0.238938
4   10.0            0.065990